### DISCUSSION TITLE
Can we accept  image and accept body as post request

### DISCUSSION THREAD

**Discussion by sankarsiva123** [fastapi/fastapi#8846] [Questions] [answered]
I tried accepting both image and body as input , but fast api throws 422(Unprocessable Entity) , can we accept both things ?

**Comment by frankie567** [comment, marked as answer]
You can't accept both a file upload and a JSON body. That's because of the HTTP protocol: when you upload a file, the encoding (content type) is `multipart/form-data`; thus it cannot be parsed by `Body` or a Pydantic model that expect JSON data. It's explained quite well in the documentation: https://fastapi.tiangolo.com/tutorial/request-files/?h=form+u#what-is-form-data

However, you can have both a file upload and a form field. You won't have all the benefits of a validation through Pydantic but it works. It looks like this:

```py
@app.post("/")
def upload_form(file: UploadFile = File(...), field1: str = Form(...), field2: str = Form(...)):
    return {"filename": file.filename, "field1": field1, "field2": field2}
```

Otherwise, what people do generally is to have two endpoints: one for the JSON data and another one exclusively for file upload.

**Comment by kvas-it** [comment]
I would like to ask the same question with a bit more detail.

I have a route like this:
```
@app.post('/data_and_files')
def data_and_files(data: DataSchema = Form(...), files: List[UploadFile] = File([])):
    # here data would be parsed DataSchema and files is a list of files.
```

FastAPI docs would say that this is not going to work "because of limitations of HTTP protocol" but FastAPI will actually render a form for submitting this request under `/docs`.  The form will produce a request like this (without the `application/json` content type in the first part, but nobody prevents us from adding that as well to make it even more kosher):

```--ecbe7a002f5cb793f8ce4897649593f6
Content-Disposition: form-data; name="data"
Content-Type: application/json

{"name": "John Doe", "age": 42}
--ecbe7a002f5cb793f8ce4897649593f6
Content-Disposition: form-data; name="files"; filename="file1.txt"
Content-Type: text/plain

content 1
--ecbe7a002f5cb793f8ce4897649593f6
Content-Disposition: form-data; name="files"; filename="file2.html"
Content-Type: text/html

<html/>
--ecbe7a002f5cb793f8ce4897649593f6--
```

FastAPI will refuse to parse the first part, because "it's not a dictionary" and I get HTTP 422. The request above is a perfectly valid HTTP request that does exactly the thing that the docs say is impossible. FastAPI could parse the `data` part into `DataSchema` as I have specified if only it realized that it should be JSON-parsed first.

For now I'm forced to change my code like this:

```
@app.post('/data_and_files')
def data_and_files(data: str = Form(...), files: List[UploadFile] = File([])):
    data = DataSchema.parse_raw(data)
    # here data would be parsed DataSchema and files is a list of files.
```

This works, but it has two disadvantages:
1.  I have to convert parsing errors into `HTTPException`s myself.
2. `/docs` shows `data` as a string, so I have to remember myself what to put into it (and when I tell my friends how to do it they complain that my API is hard to use).

All of this is not the end of the world, but it would be even better if my original code worked. Seems like FastAPI would only need to notice that it's trying to cast a string into a Pydantic model and use `DataSchema.parse_raw()` to do that. Wouldn't this be better than saying that HTTP doesn't let us do it?

**Comment by pletessier** [comment]
@frankie567, I have exactly the same problem and the way @kvas-it suggested seems to be perfectly valid. More information here by Swagger.io: https://swagger.io/docs/specification/describing-request-body/multipart-requests/
